Reorganize the project and add one plugin

Niols opened the pull request:
This PR does several things:

- it changes the build system from OCamlbuild to Dune (I personally believe it makes the build configuration much simpler, for both the main project and the plugins).

- it fixes a few "things" in order to make the project compile. Among these, the function `load_available_analyzers` has been rewriten to use only Sys (so that Unix is not needed anymore).

- it adds the `Alarm` module (moving `Analyzer.alarm` into `Alarm.t`) and adds a function to report alarms (there is room for improvement on this function, but at least it is here).

- it adds a `redirections.inverted` plugin that checks for "inverted" redirections (that is, redirections like `2>&1 >/dev/null` that do not do what is intended most of the time). This helps seeing the points mentioned earlier are indeed working.

Niols committed:
reorganize; use dune instead of ocamlbuild

Niols committed:
rewrite load_available_analyzers to avoid using Unix; the dynamic linking now works nicely

Niols committed:
fix command line parsing, add simple reporting mechanism

Niols committed:
improve alarms; report now printed correctly

Niols committed:
add an analyzer for redirections in the wrong order

yurug commented on src/plugins/redirections/crossing/checkCrossingRedirections.ml:
```diff
@@ -0,0 +1,232 @@
+open Lintshell
```
Is it the right place for these general purpose functions over CSTs? Shouldn't they moved to libmorbig?

Niols replied:
If you are talking about these `io_redirect_list_of_cmd_prefix` functions, I totally agree. That would also allow for factorization of code between shstats and lintshell in many cases.

yurug replied:
Yes.

yurug commented on src/plugins/redirections/crossing/checkCrossingRedirections.ml:
```diff
@@ -0,0 +1,232 @@
+
+(* ==================== [ Checker on redirection lists ] ==================== *)
+
+(* What we know on a descriptor *)
```
I would have inserted a comment here to explain the principle of this quite involved analysis, and its limitations.

Niols replied:
Yes, this implementation lacks documentation. Also, it seems quite overkill when we could just be detecting patterns like `2>&1 >/dev/null` specifically. While we are talking about this, do you have a better name than "crossing"? Maybe something based on "leak"?

yurug replied:
What about "inverted"?

Niols replied:
I'll take that.

yurug commented on src/plugins/redirections/crossing/checkCrossingRedirections.ml:
```diff
@@ -0,0 +1,232 @@
+               | [], _ -> check_io_redirect_list in_suffix
+               | _, _ -> check_io_redirect_list (in_prefix @ in_suffix)
+             )
+          | SimpleCommand_CmdPrefix_CmdWord (cmd_prefix', _) ->
```
There are some redundancies in the next methods implementations, right?

Niols replied:
In fact, the whole function that takes a simple command and returns a list of redirections could be externalized in Morbig's CST helpers. It makes sense as it will be used in shstats, lintshell and morsmall.

yurug replied:
Yes. (I will not be surprised to see a PR coming on the Morbig's side :)

Niols replied:
(incoming)

yurug commented:
This PR seems to do more than its title suggests.

Niols commented:
Yes, it's true. Actually, there are three things in this PR (but they are all related to each other):
- use Dune instead of OCamlbuild
- change lintshell so that it compiles (but I've done that by using the first point)
- add a non-trivial analyzer (namely, this redirection analyzer; this allowed me to see if the core of lintshell was indeed correct).

Should I rename the PR and change its description? Or try to split it into several PRs? The latter seems rather difficult, but should be doable.

yurug commented:
Updating the description and the title should be fine!

Niols committed:
move helpers about redirections in morbig

Niols committed:
rename "crossing" redirection into "inverted"

Niols committed:
document the inverted redirections checker

Niols commented:
I have just:
- added documentation to the inverted redirections checker. ("inverted" is indeed better than "crossed").
- moved the CST helpers from the inverted redirections checker into a PR to Morbig (colis-anr/morbig#36)
- changed the name and the description of this PR.

Niols committed:
follow morbig's changes for posix 2018
